src/lib/scheduler/worker.cpp

```diff
@@ -128,17 +139,55 @@ void Worker::join() {
 uint64_t Worker::num_finished_tasks() const { return _num_finished_tasks; }
 
 void Worker::_wait_for_tasks(const std::vector<std::shared_ptr<AbstractTask>>& tasks) {
-  auto tasks_completed = [&tasks]() {
-    // Reversely iterate through the list of tasks, because unfinished tasks are likely at the end of the list.
-    for (auto it = tasks.rbegin(); it != tasks.rend(); ++it) {
-      if (!(*it)->is_done()) {
+  // This lambda checks if all tasks from the vector have been executed. If they are, it causes _wait_for_tasks to
+  // return. If there are remaining tasks, it primarily tries to execute these. If they cannot be executed, the
+  // worker performs work for others (i.e., executes tasks from the queue).
+  auto check_own_tasks = [&]() {
+    auto all_done = true;
+    for (auto it = tasks.begin(); it != tasks.end(); ++it) {
+      const auto& task = *it;
+      if (task->is_done()) {
+        continue;
+      }
+
+      // Task is not yet done - check if it is ready for execution
```

Comment:
If the third task is not yet ready, but the forth is and is executed: `all_done` will be true. That's sounds strange to me.

Reply:
Correct, `all_done` will be true again, but only for a very brief moment. We reset the loop to its original state, basically starting from the front again. Once we reach the third task, it is again set to false.

Reply:
Ah, I misread the last line. I think you should at least add a comment to this kind of for loop, since it's a very unexpected behaviour for a for-loop.

Reply:
> Note: If we found a task that has not yet been executed, we reset this loop and start from the beginning. As such, both all_done and it may be reset outside of the following line.

Reply:
We don't reset, we continue. (if you want to place that comment here, where the github comment is).
But I also meant in more general on top of the for loop. Fiddling with running iterators in a for loop that looks pretty boring (begin, end, ++it), is not nice. :)

Reply:
(1) I added this at the top of the loop.
(2) We do reset, we do not just continue. I worry that having nested while/for loops with continues, outer breaks, and returns does not make things better. :/

Reply:
Yeah, I thought about while loops, but I also don't think it's any better. But a good comment should suffice though.

Reply:
Where do we reset here? Still don't see it.

Reply:
I refactored the loop, please have another look.